com.yakindu.sct.examples.coffeemachine.c/src/cm_hmi.c

```diff
@@ -4,75 +4,100 @@
  *  Created on: 13.04.2016
  *      Author: terfloth
  */
-
-#include <stdbool.h>
 #include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/select.h>
+#include <termios.h>
 
 #include "cm_hmi.h"
 
 #include "cm_trace.h"
 
+static char *choice_texts[] = { //
+		"", //
+				"coffee", //
+				"espresso", //
+				"cappucino", //
+				"latte macchiato", //
+				"milk", //
+				"" //
+		};//
+
+/*!
+ * Implementation for non blocking getchar()
+ * https://stackoverflow.com/questions/448944/c-non-blocking-keyboard-input
+ */
+struct termios orig_termios;
 
+void reset_terminal_mode() {
+	tcsetattr(0, TCSANOW, &orig_termios);
+}
 
-static UserEvents lastEvent = NONE;
+void set_conio_terminal_mode() {
+	struct termios new_termios;
 
-static char* choice_texts[] = {
-		"",
-		"coffee" ,
-		"espresso" ,
-		"cappucino" ,
-		"latte macchiato" ,
-		"milk" ,
-		""
-};
+	/* take two copies - one for now, one for later */
+	tcgetattr(0, &orig_termios);
+	memcpy(&new_termios, &orig_termios, sizeof(new_termios));
 
+	/* register cleanup handler, and set the new terminal mode */
+	atexit(reset_terminal_mode);
+	cfmakeraw(&new_termios);
+	tcsetattr(0, TCSANOW, &new_termios);
+}
 
-UserEvents provideUserEvent() {
-	UserEvents userEvent = lastEvent;
-	lastEvent = NONE;
-	return userEvent;
+int kbhit() {
+	struct timeval tv = { 0L, 0L };
+	fd_set fds;
+	FD_ZERO(&fds);
+	FD_SET(0, &fds);
+	return select(1, &fds, NULL, NULL, &tv);
 }
 
+int getch() {
+	int r;
+	unsigned char c;
+	if ((r = read(0, &c, sizeof(c))) < 0) {
+		return r;
+	} else {
+		return c;
+	}
+}
 
-void scanUserInput() {
-	while (1) {
+UserEvents getUserInput() {
+	if (kbhit()) {
```

Comment:
The tictoc tracing example uses `read()` with a prior `STDIN` configuration instead of mimic `kbhit() `. We should compare approaches.